### PR TITLE
Allow using modules as subtypes of protocols

### LINKED ISSUE #5018: Allow using modules as subtypes of protocols
There is an idea to allow modules be accepted where a protocol is expected. This pattern is sometimes used for configs and option management, for example:
```python
# file default_config.py
timeout = 100
one_flag = True
other_flag = False

# file __main__.py
import default_config
from typing import Protocol

class Options(Protocol):
    timeout: int
    one_flag: bool
    other_flag: bool

def setup(options: Options) -> None:
    ...

setup(default_config)  # OK
```
This will allow better typing than just `types.ModuleType` and should be straightforward to implement. Are there any objections against this?

### LINKED ISSUE #5439: Member is not assignable error when accessing a module member
Hi,

This issue seems close to the one that was fixed in [this PR](https://github.com/python/mypy/pull/3636).

* Are you reporting a bug, or opening a feature request?

I'm reporting a bug.

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python
import functools

class Foo:
    bar = functools

foo = Foo()
foo.bar  # error: Member "bar" is not assignable
```

* What is the actual behavior/output?

```
test.py:7: error: Member "bar" is not assignable
```

* What is the behavior/output you expect?

No Error. I'm just accessing, not mutating.

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

```
$ mypy --version
mypy 0.620
```

Same error on master.

* What are the mypy flags you are using? (For example --strict-optional)

No flags. Just raw `mypy test.py`

If you need any further information, please let me know.
Thank you in advance.

### LINKED ISSUE #10850: Final variable is mutable when using structural subtyping
**Bug Report**

Under structural subtyping with a variable narrowed to `Final`, it's possible to modify the var contents

**To Reproduce**
```
# poc.py
from dataclasses import dataclass
from typing import Protocol, Final

@dataclass
class FinalY:
    y: Final[int] = 0

class Y(Protocol):
    y: int

def f(x: Y) -> Y:
    x.y = 42
    return x

print(f(FinalY(17)))
```

>$ mypy poc.py  
>Success: no issues found in 1 source file

>$ python poc.py
>FinalY(y=42)

**Expected Behavior**

Class with a variable `Final[T]` should not be a structural subtype of a protocol with variable `T`

**Your Environment**

- Mypy version used: 0.910
- Python version used: 3.9.6

_This bug was inspired by @erictraut's [example of how to deal with other kind of issue](https://github.com/python/mypy/issues/5374#issuecomment-884220116)_

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes #5018 
Fixes #5439
Fixes #10850 

The implementation is simple but not the most beautiful one. I simply add a new slot to the `Instance` class that represents content of the module. This new attribute is short lived (it is not serialized, and not even stored on variables etc., because we erase it in `copy_modified()`). We don't need to store it, because all the information we need is already available in `MypyFile` node. We just need the new attribute to communicate between the checker and `subtypes.py`.

Other possible alternatives like introducing new dedicated `ModuleType`, or passing the symbol tables to `subtypes.py` both look way to complicated. Another argument in favor of this new slot is it could be useful for other things, like `hasattr()` support and ad hoc callable attributes (btw I am already working on the former).

Note there is one important limitation: since we don't store the module information, we can't support module objects stored in nested positions, like `self.mods = (foo, bar)` and then `accepts_protocol(self.mods[0])`. We only support variables (name expressions) and direct instance, class, or module attributes (see tests). I think this will cover 99% of possible use-cases.

### CODE REVIEW

**Comment by ilevkivskyi** [comment]
Hm, those few errors are clearly wrong, and I know where they are coming from. Will try to fix now.

File: test-data/unit/check-protocols.test [hunk 1]
```diff
@@ -3546,3 +3546,142 @@ S = TypeVar("S")
...
+other_flag = False
+def update() -> str: ...
+[builtins fixtures/module.pyi]
+
+[case testModuleAsProtocolImplementationClassObject]
+import runner
+import bad_runner
+from typing import Callable, Protocol
+
+class Runner(Protocol):
```
Comment (sobolevn) [inline test-data/unit/check-protocols.test:3628]: What if some protocol requires `ClassVar`?

```python
class My(Protocol):
    A: ClassVar[int]
```

Can this be satisfied by a module (I think not)? Do we need a test for this?

File: test-data/unit/check-protocols.test [hunk 2]
```diff
@@ -3546,3 +3546,142 @@ S = TypeVar("S")
...
+def run(x: Runner) -> None: ...
+run(runner)  # OK
+run(bad_runner)  # E: Argument 1 to "run" has incompatible type Module; expected "Runner" \
+                 # N: Following member(s) of "Module bad_runner" have conflicts: \
+                 # N:     Expected: \
+                 # N:         def (int, /) -> Result \
+                 # N:     Got: \
+                 # N:         def __init__(arg: str) -> Run
+
+[file runner.py]
```
Comment (sobolevn) [inline test-data/unit/check-protocols.test:3644]: I think it makes sence to test something like:

```python
from typing_extensions import Protocol
class My(Protocol):
   a: int  # is settable attr

def func(arg: My): pass

import some_module
func(some_module)  # error
```

And:

```python
# some_module.py
from typing_extensions import Final

a: Final = 1  # is module-level constant
```

Looks like we should correctly raise an error here.

**Comment by ilevkivskyi** [comment]
Yes, I think both should be errors. I will add tests later today (and update implementation if needed).

**Comment by ilevkivskyi** [comment]
@sobolevn OK, I added test and fixed code. Btw I discovered that we didn't do the final check for normal instances either, so I fixed that as well (I actually even found an issue about this on the tracker).

**Review by sobolevn — APPROVED** [review]
🎉

**Comment by gvanrossum** [comment]
You rock! Also thank you to the reviewers.

**Comment by ilevkivskyi** [comment]
Got some free time on my paternity leave :-)

**Comment by gvanrossum** [comment]
> Got some free time on my paternity leave :-)

Oh congratulations!

**Comment by ilevkivskyi** [comment]
Thank you!

**Pull request merged**